src/flist/db.rs

```diff
@@ -0,0 +1,135 @@
+use anyhow::{bail, Result};
+use futures::StreamExt;
+use md5::{Digest, Md5};
+use std::path::{Path, PathBuf};
+use tokio::fs;
+use tokio::io::{self, AsyncWriteExt};
+use uuid::Uuid;
+pub struct MetadataDbMgr {
+    // root directory where all
+    // the working file of the module will be located
+
+    // underneath are the path for each
+    // sub folder used by the flist module
+    flist: PathBuf,
+}
+
+impl MetadataDbMgr {
+    pub async fn new<P: AsRef<Path>>(flist: P) -> Result<Self> {
+        Ok(Self {
+            flist: flist.as_ref().into(),
+        })
+    }
+
+    pub async fn get<T: AsRef<str>>(&self, url: T) -> Result<PathBuf> {
+        let url = url.as_ref();
+        let hash = self.hash_of_flist(url).await?;
+        let path = self.flist.join(&hash);
+        match fs::File::open(&path).await {
```

Comment:
If you actually don't use the `opened` file. may be do `fs::metadata()` instead

Reply:
A better way to do it is actually "use" the opened file. Since download will still have to open the file but you already have it open, you can open the file in read/write mode. If it's after open of a zero size (just created) then you can pass this open file to the download function to "fill it up" with the downloaded data.

If it exists and is not zero size, then you can compare it's hash and redownload if needed